### PR TITLE
Tutorial for DDP + RPC.

### PR DESCRIPTION
**Pull request by pritamdamania87**
Summary: Based on example from https://github.com/pytorch/examples/pull/800

### CODE REVIEW

**Comment by ezyang** [comment]
It would be much better if the tutorial code was written in runnable fashion (either by writing the tutorial directly as a python file with sphinx-gallery, or keeping the Python code separate from the rst file) so that we can test the tutorial in the future. See https://github.com/pytorch/tutorials/pull/1057 for an example of how to do this.

File: advanced_source/rpc_ddp_tutorial.rst [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+In this tutorial we will cover case 1 mentioned above. We have a total of 4 
+workers in our setup as follows:
+
+
+1) 1 Master, which is responsible for creating an embedding table 
+   (nn.EmbeddingBag) on the parameter server. The master also drives the 
+   training loop on the two trainers.
+2) 1 Parameter Server, which basically holds the embedding table in memory and 
+   responds to RPCs from the Master and Trainers.
+3) 2 Trainers, which store a FC layer (nn.Linear) which is replicated amongst
```
Comment (mrshenli) [inline advanced_source/rpc_ddp_tutorial.rst:39]: an FC

File: advanced_source/rpc_ddp_tutorial.rst [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+   responds to RPCs from the Master and Trainers.
+3) 2 Trainers, which store a FC layer (nn.Linear) which is replicated amongst 
+   themselves using `DistributedDataParallel <https://pytorch.org/docs/stable/nn.html#torch.nn.parallel.DistributedDataParallel>`__.
+   The trainers are also responsible for executing the forward pass, backward 
+   pass and optimizer step.
+
+|
+The entire training process is executed as follows:
+
+1) The master creates an embedding table on the Parameter Server and holds a
```
Comment (mrshenli) [inline advanced_source/rpc_ddp_tutorial.rst:47]: holds an

File: advanced_source/rpc_ddp_tutorial.rst [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+   themselves using `DistributedDataParallel <https://pytorch.org/docs/stable/nn.html#torch.nn.parallel.DistributedDataParallel>`__.
+   The trainers are also responsible for executing the forward pass, backward 
+   pass and optimizer step.
+
+|
+The entire training process is executed as follows:
+
+1) The master creates an embedding table on the Parameter Server and holds a 
+   `RRef <https://pytorch.org/docs/master/rpc.html#rref>`__ to it.
+2) The master, then kicks of the training loop on the trainers and passes the
```
Comment (mrshenli) [inline advanced_source/rpc_ddp_tutorial.rst:49]: kicks off

File: advanced_source/rpc_ddp_tutorial.rst [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+6) Next, Distributed Autograd propagates the gradients to the parameter server, 
+   where the gradients for the embedding table are updated.
+7) Finally, the `Distributed Optimizer <https://pytorch.org/docs/master/rpc.html#module-torch.distributed.optim>`__ is used to update all the parameters.
+
+
+|
+**NOTE**: You should always use `Distributed Autograd <https://pytorch.org/docs/master/rpc.html#distributed-autograd-framework>`__ for the backward pass if you're combining DDP and RPC.
+
+
+Now, lets go through each part in detail. Firstly, we need to setup all of our
```
Comment (mrshenli) [inline advanced_source/rpc_ddp_tutorial.rst:67]: let's

File: advanced_source/rpc_ddp_tutorial.rst [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+Once RPC initialization is done, the master creates an `EmbeddingBag <https://pytorch.org/docs/master/generated/torch.nn.EmbeddingBag.html>`__ 
+on the Parameter Server using `rpc.remote <https://pytorch.org/docs/master/rpc.html#torch.distributed.rpc.remote>`__.
+The master then loops through each trainer and kicks of the training loop by 
+calling ``_run_trainer`` on each trainer using `rpc_async <https://pytorch.org/docs/master/rpc.html#torch.distributed.rpc.rpc_async>`__.
+Finally, the master waits for all training to finish before exiting.
+
+The trainers first initialize a ``ProcessGroup`` for DDP with world_size=2 
+(for two trainers) using `init_process_group <https://pytorch.org/docs/stable/distributed.html#torch.distributed.init_process_group>`__.
+Next, they initialize the RPC framework using the TCP init_method. Note that 
+the ports are different in RPC initialization and ProcessGroup intialization.
```
Comment (mrshenli) [inline advanced_source/rpc_ddp_tutorial.rst:82]: intialization -> initialization

File: index.rst [hunk 6]
```diff
@@ -332,6 +332,13 @@ Welcome to PyTorch Tutorials
    :link: intermediate/rpc_param_server_tutorial.html
    :tags: Parallel-and-Distributed-Training
 
+.. customcarditem::
```
Comment (mrshenli) [inline index.rst:335]: We also need to add an item to the left navigation bar, which is at the end of this file (search for `:caption: Parallel and Distributed Training`).

File: advanced_source/rpc_ddp_tutorial.rst [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+        # block until all rpcs finish
+        rpc.shutdown()
+
+
+    if __name__=="__main__":
+        # 2 trainers, 1 parameter server, 1 master.
+        world_size = 4
+        mp.spawn(run_worker, args=(world_size, ), nprocs=world_size, join=True)
+
+Before we discuss details of the Trainer, lets introduce the ``HybridModel`` that
```
Comment (mrshenli) [inline advanced_source/rpc_ddp_tutorial.rst:162]: let's

File: advanced_source/rpc_ddp_tutorial.rst [hunk 8]
```diff
@@ -0,0 +1,308 @@
...
+
+    if __name__=="__main__":
+        # 2 trainers, 1 parameter server, 1 master.
+        world_size = 4
+        mp.spawn(run_worker, args=(world_size, ), nprocs=world_size, join=True)
+
+Before we discuss details of the Trainer, lets introduce the ``HybridModel`` that 
+the trainer uses. As described below, the ``HybridModel`` is initialized using an 
+RRef to the embedding table (emb_rref) on the parameter server and the ``device`` 
+to use for DDP. The initialization of the model wraps a
```
Comment (mrshenli) [inline advanced_source/rpc_ddp_tutorial.rst:165]: a -> an

File: advanced_source/rpc_ddp_tutorial.rst [hunk 9]
```diff
@@ -0,0 +1,308 @@
...
+          super(HybridModel, self).__init__()
+          self.emb_rref = emb_rref
+          self.fc = DDP(torch.nn.Linear(16, 8).cuda(device), device_ids=[device])
+          self.device = device
+
+      def forward(self, indices, offsets):
+          emb_lookup = self.emb_rref.rpc_sync().forward(indices, offsets)
+          return self.fc(emb_lookup.cuda(self.device))
+
+Next, lets look at the setup on the Trainer. The trainer first creates the
```
Comment (mrshenli) [inline advanced_source/rpc_ddp_tutorial.rst:198]: let's

**Comment by mrshenli** [comment]
cc @jlin27

**Comment by pritamdamania87** [comment]
> IIRC, tutorials for v1.6 need to use release/1.6 as the target branch.

Don't we need to commit the tutorial in master first though?

@ezyang Thanks for the great suggestion! I used `literalinclude` and this approach is much better.

@mrshenli @jlin27 Do we still need to add examples in https://github.com/pytorch/examples/pull/800 if we have a runnable example in the tutorials repo? It seems like we can keep the runnable example in the tutorials repo and just refer to it in our tutorial?

**Comment by jlin27** [comment]
> > IIRC, tutorials for v1.6 need to use release/1.6 as the target branch.
> 
> Don't we need to commit the tutorial in master first though?
The pytorch/tutorials branches are backwards from pytorch/pytorch. In pytorch/tutorials, `master` is live/stable (not sure the history but it is what it is rn :) ) For tutorials planned for 1.6, please submit them to the `release/1.6` branch and I'll get them live on master for the release. 

> @mrshenli @jlin27 Do we still need to add examples in [pytorch/examples#800](https://github.com/pytorch/examples/pull/800) if we have a runnable example in the tutorials repo? It seems like we can keep the runnable example in the tutorials repo and just refer to it in our tutorial?

I think the pro of keeping it in the examples as well is that there are a number of other related examples there already and it would be advantageous for users to find this one there. Only downside is maintaining both.

**Comment by pritamdamania87** [comment]
Created a PR against release/1.6: https://github.com/pytorch/tutorials/pull/1071

**Comment by mrshenli** [comment]
> @mrshenli @jlin27 Do we still need to add examples in pytorch/examples#800 if we have a runnable example in the tutorials repo? It seems like we can keep the runnable example in the tutorials repo and just refer to it in our tutorial?

I am not sure about the plan here. @ezyang do you know if we are going to completely get rid the `pytorch/examples` repo and stick all source code in the `pytorch/tutorials` repo (I assume all examples should have their associated tutorial)? 

If this is the plan, then yes, let's move all distributed examples source code to tutorials repo. 

If this is **not** the plan, then what is our guidelines to users regarding what to expect in `pytorch/examples` vs `pytorch/tutorials`? Looks like both repos will become some random subset of the example projects we have.

Another question for @jlin27 is we do plan to let tutorials CI to run the source code included in tutorials? Or will this be test be installed to the examples repo?

**Comment by jlin27** [comment]
> Another question for @jlin27 is we do plan to let tutorials CI to run the source code included in tutorials? Or will this be test be installed to the examples repo?

The plan is to have it on both examples and tutorials. I haven't heard anything about getting rid of `pytorch/examples`, but please do comment if that's been under discussion.

**Comment by ezyang** [comment]
I don't think it's a good idea to have two copies of the code. Perhaps an easy compromise is to submodule examples inside tutorials. Then you can literalinclude the examples code directly.

**Pull request closed without merging**